### PR TITLE
[Deepspeed] add support for bf16 mode

### PR DESCRIPTION
**Pull request by stas00**
This PR:

- adds support for bfloat16 for ZeRO-1, ZeRO-2 and ZeRO-3 to HF/DS integration.
- most functional tests are now run for bf16 as well - so this PR almost doubles the number of tests. model zoo tests are left with fp16 for now, as both should work the same.
- docs are updated to document the new feature

Requirements:

1. [x] merged https://github.com/huggingface/transformers/pull/13207
2. [x] ZeRO-3 support merged https://github.com/microsoft/DeepSpeed/pull/1453
3. [x] Need new deepspeed release after the above is merged 
4. [x] update version dependency table

---------------------------

for users who want to try this early:

1. add `--bf16` to your previous fp16 or fp32 deepspeed command line 

done ;)

@sgugger

### CODE REVIEW

File: docs/source/main_classes/deepspeed.mdx [hunk 1]
```diff
@@ -1248,7 +1251,7 @@ To configure pytorch AMP-like mode set:
 and the [`Trainer`] will automatically enable or disable it based on the value of
 `args.fp16_backend`. The rest of config values are up to you.
 
-This mode gets enabled when `--fp16 --fp16_backend amp` command line args are passed.
+This mode gets enabled when ``--fp16 --fp16_backend amp`` or ``--fp16_full_eval`` command line args are passed.
```
Comment (sgugger) [inline docs/source/main_classes/deepspeed.mdx:1254]:
```suggestion
This mode gets enabled when `--fp16 --fp16_backend amp` or `--fp16_full_eval` command line args are passed.
```
Why the double back-ticks?

File: docs/source/main_classes/deepspeed.mdx [hunk 2]
```diff
@@ -1270,6 +1273,35 @@ configuration.
...
+{
+    "bf16": {
+        "enabled": "auto"
+    }
+}
+```
+
+bf16 has the same dynamic range as fp32 and thus doesn't require loss scaling.
+
+This mode gets enabled when ``--bf16`` or ``--bf16_full_eval`` command line args are passed.
```
Comment (sgugger) [inline docs/source/main_classes/deepspeed.mdx:1290]: Same here?

File: src/transformers/deepspeed.py [hunk 3]
```diff
@@ -228,26 +230,39 @@ def trainer_config_process(self, args):
...
             self._dtype = torch.float32
-        # later there will be other dtypes besides just fp16 and fp32
-        # also not quite sure what dtype should be under apex, defaulting to fp16 for now
+        else:
+            self._dtype = torch.float16
+
+        # print(self._dtype)
+        # print(self.get_value("fp16.enabled"))
+        # print(self.get_value("bf16.enabled"))
+        # print(self.config)
```
Comment (sgugger) [inline src/transformers/deepspeed.py:264]: Still not a big fan of those comments.

File: tests/deepspeed/test_deepspeed.py [hunk 4]
```diff
@@ -218,14 +238,15 @@ def setUp(self):
         with io.open(self.ds_config_file[ZERO2], "r", encoding="utf-8") as f:
             config_zero2 = json.load(f)
             # by default use fp16
-            config_zero2["fp16"]["enabled"] = True
+            # config_zero2["fp16"]["enabled"] = True
```
Comment (sgugger) [inline tests/deepspeed/test_deepspeed.py:241]: Why leave it commented out? I understand the print statements are for quick debugging, but this should be cleaned up, no?

File: tests/deepspeed/test_deepspeed.py [hunk 5]
```diff
@@ -218,14 +238,15 @@ def setUp(self):
...
         with io.open(self.ds_config_file[ZERO3], "r", encoding="utf-8") as f:
             config_zero3 = json.load(f)
             # by default use fp16
-            config_zero3["fp16"]["enabled"] = True
+            # config_zero3["fp16"]["enabled"] = True
```
Comment (sgugger) [inline tests/deepspeed/test_deepspeed.py:245]: Same here.

File: tests/deepspeed/test_deepspeed.py [hunk 6]
```diff
@@ -510,37 +530,42 @@ def check_saved_checkpoints_deepspeed(self, output_dir, freq, total, stage):
...
         freq = 5
         output_dir = self.get_auto_remove_tmp_dir()
         ds_config_dict = self.get_config_dict(stage)
-        ds_config_dict["fp16"]["initial_scale_power"] = 1  # force optimizer on the first step
-        if stage == ZERO3:
-            ds_config_dict["zero_optimization"]["stage3_gather_fp16_weights_on_model_save"] = True
+        if dtype == FP16:
+            ds_config_dict["fp16"]["initial_scale_power"] = 1  # force optimizer on the first step
+        # XXX:
+        if stage == ZERO3:  # and dtype == FP16:
```
Comment (sgugger) [inline tests/deepspeed/test_deepspeed.py:543]: I don't understand those comments.

File: tests/deepspeed/test_deepspeed.py [hunk 7]
```diff
@@ -558,19 +583,20 @@ def test_can_resume_training_errors(self, stage):
...
-        output_dir = self.get_auto_remove_tmp_dir()
+        output_dir = self.get_auto_remove_tmp_dir("./xxx", after=False)
         ds_config_dict = self.get_config_dict(stage)
-        ds_config_dict["fp16"]["initial_scale_power"] = 1  # force optimizer on the first step
-        if stage == ZERO3:
-            ds_config_dict["zero_optimization"]["stage3_gather_fp16_weights_on_model_save"] = True
+        if dtype == FP16:
+            ds_config_dict["fp16"]["initial_scale_power"] = 1  # force optimizer on the first step
+        # XXX:
+        if stage == ZERO3:  # and dtype == FP16:
```
Comment (sgugger) [inline tests/deepspeed/test_deepspeed.py:595]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: src/transformers/deepspeed.py (hunk 3 above)
Comment (stas00) [reply src/transformers/deepspeed.py:264]: sorry, this wasn't intended to leave behind - will clean up

File: tests/deepspeed/test_deepspeed.py (hunk 4 above)
Comment (stas00) [reply tests/deepspeed/test_deepspeed.py:241]: thank you for catching - cleaned up!

File: tests/deepspeed/test_deepspeed.py (hunk 6 above)
Comment (stas00) [reply tests/deepspeed/test_deepspeed.py]: oops, was waiting for last function to be implemented in deespeed - all cleaned up now.

**Comment by stas00** [comment]
Thanks a lot for catching those few development process leftovers, Sylvain! Much appreciating. All cleaned up now.

**Review by sgugger — APPROVED** [review]
Looks all good to me!

**Pull request merged**